### PR TITLE
Img verificator

### PR DESCRIPTION
**Pull request by ggruszczynski**

### CODE REVIEW

File: apps/blender/task/blenderrendertask.py [hunk 1]
```diff
@@ -485,7 +488,9 @@ def query_extra_data_for_test_task(self):
 
         hash = "{}".format(random.getrandbits(128))
 
-        self.test_task_res_path = get_test_task_path(self.root_path)
+        dm = DirManager(self.root_path)
+        self.test_task_res_path = dm.get_task_test_dir(self.header.task_id) #get_test_task_path(self.root_path)
```
Comment (badb) [inline apps/blender/task/blenderrendertask.py]: Remove the commented line, please

File: apps/blender/task/blenderrendertask.py [hunk 2]
```diff
@@ -11,7 +11,10 @@
 
 from golem.core.common import to_unicode
 from golem.core.fileshelper import has_ext
-from golem.resource.dirmanager import get_test_task_path
+
+from golem.resource.dirmanager import DirManager
+
+
```
Comment (badb) [inline apps/blender/task/blenderrendertask.py]: Unnecessary double extra line

File: apps/blender/task/verificator.py [hunk 3]
```diff
@@ -85,6 +85,4 @@ def _get_part_size_from_subtask_number(self, subtask_number):
             else:
                 res_y = ceiling_height
         return res_y
-
-    def _check_size(self, file_, res_x, res_y):
```
Comment (badb) [inline apps/blender/task/verificator.py]: Why do you remove this function?

File: apps/core/task/verificator.py [hunk 4]
```diff
@@ -36,6 +36,7 @@ def set_verification_options(self, verification_options):
         else:
             self.advanced_verification = True
 
+
```
Comment (badb) [inline apps/core/task/verificator.py]: Not needed.

File: apps/lux/task/luxrendertask.py [hunk 5]
```diff
@@ -147,15 +151,40 @@ def __init__(self, halttime, haltspp, **kwargs):
...
 
+        self.random_crop_window_for_verification = \
+            self._get_random_crop_window_for_verification(self.scene_file_src)
+
         self.output_file, _ = os.path.splitext(self.output_file)
         self.output_format = self.output_format.lower()
         self.num_add = 0
 
         self.preview_exr = None
+        self.referenceRuns = 2
```
Comment (badb) [inline apps/lux/task/luxrendertask.py]: https://www.python.org/dev/peps/pep-0008/#method-names-and-instance-variables Please keep isntance variables lowercase with underscore. Also consider making magic numer 2 constant

File: apps/lux/task/luxrendertask.py [hunk 6]
```diff
@@ -147,15 +151,40 @@ def __init__(self, halttime, haltspp, **kwargs):
...
+
+    def _get_random_crop_window_for_verification(self,
```
Comment (badb) [inline apps/lux/task/luxrendertask.py]: Consider moving this function to scenefilereader

File: apps/lux/task/luxrendertask.py [hunk 7]
```diff
@@ -208,6 +243,7 @@ def query_extra_data(
             self.output_format
         )
         scene_dir = os.path.dirname(self._get_scene_file_rel_path())
+        # self._temp_save("provscenejob.txt", scene_src)
```
Comment (badb) [inline apps/lux/task/luxrendertask.py]: should be removed?

File: apps/lux/task/verificator.py [hunk 8]
```diff
@@ -2,44 +2,129 @@
...
 from apps.core.task.verificator import SubtaskVerificationState
 from apps.rendering.task.verificator import RenderingVerificator
 
+from apps.rendering.resources.ImgVerificator import ImgStatistics, ImgVerificator
+from apps.rendering.resources.imgrepr import (PILImgRepr)
 
 logger = logging.getLogger("apps.lux")
 
 
+
```
Comment (badb) [inline apps/lux/task/verificator.py]: Not needed

File: apps/lux/task/verificator.py [hunk 9]
```diff
@@ -2,44 +2,129 @@
...
+            f = glob.glob(os.path.join(dir,'*.png'))
+
+            ref_img = PILImgRepr()
+            ref_img.load_from_file(f.pop())
+            ref_imgs.append(ref_img)
+
+        return ref_imgs
+
+    # def _advanced_check_files(self, tr_png_files, tr_flm_files):
```
Comment (badb) [inline apps/lux/task/verificator.py]: Remove?

File: apps/lux/task/verificator.py [hunk 10]
```diff
@@ -2,44 +2,129 @@
...
+
+
+        if self.advanced_verification:
+            self.test_flm = self._get_test_flm(task)
+            if not os.path.isfile(self.test_flm):
+                logger.warning("Advanced verification set, but couldn't find flm for merging test!")
+                logger.warning("No merge verification")
+
+
+            imgVerificator = ImgVerificator()
```
Comment (badb) [inline apps/lux/task/verificator.py]: img_verificator, please.

File: apps/lux/task/verificator.py [hunk 11]
```diff
@@ -2,44 +2,129 @@
...
 import shutil
+import glob
 
 from golem.core.fileshelper import common_dir, find_file_with_ext, has_ext
 from golem.task.localcomputer import LocalComputer
 
 from apps.core.task.verificator import SubtaskVerificationState
 from apps.rendering.task.verificator import RenderingVerificator
 
+from apps.rendering.resources.ImgVerificator import ImgStatistics, ImgVerificator
```
Comment (badb) [inline apps/lux/task/verificator.py]: Try to keep lines no longer then 80 characters. This applies to all changes in this file.

File: apps/rendering/resources/ImgVerificator.py [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+from apps.rendering.resources.imgrepr import (ImgRepr, PILImgRepr)
+from apps.core.task.verificator \
+    import SubtaskVerificationState as VerificationState
+
+from ssim import compute_ssim
+
+logger = logging.getLogger("apps.rendering")
+
+
+class ImgStatistics:
```
Comment (badb) [inline apps/rendering/resources/ImgVerificator.py]: Use new style classes, please.

File: apps/rendering/resources/ImgVerificator.py [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+        if imgStat.ssim > maybe_ratio * reference_imgStat.ssim: \
+                # and acceptance_ratio * imgStat.mse_bw
+            # < reference_imgStat.mse_bw:
+            # and imgStat.psnr > acceptance_ratio
+            #  * reference_imgStat.psnr:
+            return VerificationState.UNKNOWN
+
+        return VerificationState.WRONG_ANSWER
+
+    def find_outliers(self, imgStats):
```
Comment (badb) [inline apps/rendering/resources/ImgVerificator.py]: Do you need it?

File: apps/rendering/resources/imgcompare.py [hunk 14]
```diff
@@ -21,16 +21,27 @@ def calculate_psnr(mse, max_=255):
         raise ValueError("MSE & MAX_ must be higher than 0")
     return 20 * math.log10(max_) - 10 * math.log10(mse)
 
-
```
Comment (badb) [inline apps/rendering/resources/imgcompare.py]: .... and this is needed.

File: apps/rendering/resources/imgcompare.py [hunk 15]
```diff
@@ -41,10 +52,14 @@ def calculate_mse(img1, img2, start1=(0, 0), start2=(0, 0), box=None):
 
     if res_x <= 0 or res_y <= 0:
         raise ValueError("Image or box resolution must be greater than 0")
+
     mse /= res_x * res_y * 3
     return mse
 
 
+
+
```
Comment (badb) [inline apps/rendering/resources/imgcompare.py]: Too many empty lines

File: apps/rendering/resources/imgcompare.py [hunk 16]
```diff
@@ -82,6 +97,8 @@ def compare_exr_imgs(file1, file2):
         return False
 
 
+
```
Comment (badb) [inline apps/rendering/resources/imgcompare.py]: This is not needed:

File: apps/rendering/resources/imgcompare.py [hunk 17]
```diff
@@ -102,7 +119,7 @@ def _box_too_big(box):
             logger.error("Wrong box size for advanced verification " \
                          "{}".format(box_size))
 
-        if isinstance(img, PILImgRepr) and isinstance(cmp_img, PILImgRepr):
+        if isinstance(img, PILImgRepr) and isinstance(cmp_img, PILImgRepr): #
```
Comment (badb) [inline apps/rendering/resources/imgcompare.py]: Remove "#"

File: apps/rendering/task/verificator.py [hunk 18]
```diff
@@ -46,17 +46,14 @@ def _verify_imgs(self, subtask_id, subtask_info, tr_files, task):
 
         file_for_adv_ver = self._choose_adv_ver_file(tr_files, subtask_info)
         if file_for_adv_ver:
-            if not self.make_advance_verification(file_for_adv_ver,
-                                                  subtask_info, subtask_id,
-                                                  task):
+            if not self.make_advance_verification(file_for_adv_ver, subtask_info, subtask_id, task):
```
Comment (badb) [inline apps/rendering/task/verificator.py]: Under 80, please!

File: golem/task/localcomputer.py [hunk 19]
```diff
@@ -27,7 +27,7 @@ def __init__(self, task, root_path, success_callback, error_callback, get_comput
         self.success = False
         self.lock = Lock()
         self.tt = None
-        self.root_path = root_path
+        self.dirManager = DirManager(root_path)
```
Comment (badb) [inline golem/task/localcomputer.py]: dir_manager

File: golem/task/localcomputer.py [hunk 20]
```diff
@@ -101,16 +101,18 @@ def _get_time_spent(self):
 
     def __prepare_resources(self):
 
-        self.test_task_res_path = get_test_task_path(self.root_path)
+        self.test_task_res_path = self.dirManager.get_task_test_dir("")#self.task.header.task_id)
```
Comment (badb) [inline golem/task/localcomputer.py]: some cleaning up would be nice

File: golem/task/taskmanager.py [hunk 21]
```diff
@@ -168,11 +171,14 @@ def add_new_task(self, task):
 
         self.tasks_states[task.header.task_id] = ts
 
+
```
Comment (badb) [inline golem/task/taskmanager.py]: Not needed (same for lines below)

File: tests/apps/rendering/resources/test_imgcompare.py [hunk 22]
```diff
@@ -3,12 +3,15 @@
 from PIL import Image
 
 
-from apps.rendering.resources.imgcompare import (advance_verify_img,
-                                                 check_size, compare_exr_imgs,
-                                                 compare_imgs,
-                                                 compare_pil_imgs,
-                                                 calculate_mse,
-                                                 calculate_psnr, logger)
+# from apps.rendering.resources.imgcompare import (advance_verify_img,
```
Comment (badb) [inline tests/apps/rendering/resources/test_imgcompare.py]: Just remove it.

**Review by badb — CHANGES_REQUESTED** [review]
Some issues about style

File: apps/blender/task/verificator.py (hunk 3 above)
Comment (ggruszczynski) [inline apps/blender/task/verificator.py]: The override does exactly the same as base function.

Comment (ggruszczynski) [reply apps/blender/task/verificator.py]: its is back now

**Comment by badb** [comment]
It doesn't work for .exr output files. Reference image is always generated with png extension, but final_image_ready checks for default output format.

**Comment by badb** [comment]
Can  you bring back protocol versions? Also please, don't touch reputation in this branch.

File: apps/lux/task/verificator.py [hunk 23]
```diff
@@ -61,84 +45,115 @@ def _get_reference_imgs(self, task):
...
                 dm.get_ref_data_dir(task.header.task_id, counter=i),
                 dm.tmp,
-                dm.output
-                )
+                dm.output)
+
+            f = glob.glob(os.path.join(dir, '*.' + task.output_format))
 
-            f = glob.glob(os.path.join(dir, '*.png'))
+            if task.output_format == "exr":
```
Comment (badb) [inline apps/lux/task/verificator.py]: You can use load_img function from imgrepr

**Pull request merged**